src/main/java/com/orange/fiware/openlpwa/iotagent/Agent.java

```diff
@@ -182,6 +191,34 @@ public void newMessageArrived(String deviceID, DeviceIncomingMessage incomingMes
 
             if (converter != null) {
                 List<ContextAttribute> decodedAttributes = converter.decodeData(deviceID, payload, incomingMessage);
+                try {
+                    ngsiManager.updateDeviceAttributes(deviceID, decodedAttributes).addCallback(
+                            updateContextResponse -> {
+                                if (updateContextResponse != null && updateContextResponse.getErrorCode().getCode().equals("200")) {
+                                    logger.info("Message sent successfully.");
+                                }
+                            },
+                            ex -> {
+                                logger.error("An error occurred while sending the message: error{}", ex.getMessage());
+                                if (ex.getMessage().contains("401")) {
+                                    logger.info("Try sending the message again with an updated token.");
+                                    try {
+                                        ngsiManager.setAccessTokenSync();
+                                        ngsiManager.updateDeviceAttributes(deviceID, decodedAttributes).addCallback(
+                                                updateContextResponse -> {
+                                                    if (updateContextResponse != null && updateContextResponse.getErrorCode().getCode().equals("200")) {
+                                                        logger.info("Message sent successfully.");
+                                                    }
```

Comment:
Any error (400, 401, etc.) is treated as an exception and logged as such :` e -> logger.error("An error occurred while sending the message: error{}", e.getMessage()));`

I will add logging in case of a null response though.